### ISSUE TITLE
Getting the log of a VM instance

### ISSUE THREAD

**Issue by gberginc** [Mirantis/virtlet#249]
Last week we discussed with @miha-plesko about a possibility of getting complete log of a virtual machine launched by Virtlet (libvirt). It was mainly focusing on the unikernels because there is frequently a need to analyse the whole log to debug the problems.

I went on to check what exactly OpenStack does when creating libvirt domains (stripped down version):

```xml
    <serial type='file'>
      <source path='/tmp/osv-cli.console.log'/>
      <target port='0'/>
    </serial>
    <serial type='pty'>
      <target port='1'/>
    </serial>
    <console type='file'>
      <source path='/tmp/osv-cli.console.log'/>
      <target type='serial' port='0'/>
    </console>
```

This will create a log file at the given location and write everything the VM writes during boot and runtime (including application log messages).

The problem with this approach is that `virsh console` will no longer work. It's possible to use `virsh console --devname=serial1`, but this will not accept any input.

In my opinion having an easy to find log file would be preferable to supporting `virsh console`, but I can imagine use-cases where the console might be beneficial.

Thus, @miha-plesko, can you please try to alter the libvirt domain [template](https://github.com/Mirantis/virtlet/blob/2f7a3a1fc94122e978d9058a642470b945c8c80d/pkg/libvirttools/virtualization.go#L198-L203) to include the above and see if you get the log file (please note that in this case, you don't actually need the second `serial type="pty"`.

In my case, I get the following (boot, followed by `ls -al` and `ls /dev`):

```
lemmy@mike-dev:~/dev/osv⟫ sudo cat /tmp/osv-cli.console.log
OSv v0.24-355-g97bbc87
/# ls -al
total 5415
drwxrwxrwx 10 osv osv 20      Apr 10 20:43 .
drwxrwxrwx 10 osv osv 20      Apr 10 20:43 ..
drwxrwxrwx 4  osv osv 6       Apr 10 20:43 cli
drwx       0  osv osv 0       Jan 01 1970  dev
drwxrwxrwx 2  osv osv 6       Apr 10 20:43 etc
drwxrwxrwx 2  osv osv 3       Apr 10 20:43 init
-rwxrwxrwx 1  osv osv 97064   Apr 10 20:43 libboost_filesystem.so.1.58.0
-rwxrwxrwx 1  osv osv 1082648 Apr 10 20:43 libboost_regex.so.1.58.0
-rwxrwxrwx 1  osv osv 14496   Apr 10 20:43 libboost_system.so.1.58.0
-rwxrwxrwx 1  osv osv 10280   Apr 10 20:43 libenviron.so
-rwxrwxrwx 1  osv osv 1087648 Apr 10 20:43 libhttpserver.so
-rwxrwxrwx 1  osv osv 23240   Apr 10 20:43 libtools.so
-rwxrwxrwx 1  osv osv 39952   Apr 10 20:43 libuutil.so
-rwxrwxrwx 1  osv osv 212168  Apr 10 20:43 libzfs.so
drwx       0  osv osv 0       Jan 01 1970  proc
drwxrwxrwx 2  osv osv 2       Apr 10 20:43 tmp
drwxrwxrwx 2  osv osv 6       Apr 10 20:43 tools
drwxrwxrwx 6  osv osv 6       Apr 10 20:43 usr
-rwxrwxrwx 1  osv osv 99184   Apr 10 20:43 zfs.so
-rwxrwxrwx 1  osv osv 98592   Apr 10 20:43 zpool.so
/# ls /dev
console null    random  urandom vblk0   vblk0.1 zfs
```

**Comment by jellonek** [comment]
But what above using http://www.conserver.com/ as a middleware which also supports file type output (which then can be later reused for PTY and file type output) ?
This was mentioned in previous talk and hopefully will cover wider aspect of cases.

So - I see this as conserv or similar approach (possibly coded in golang, but imo conserver license is good - http://www.conserver.com/LICENSE - i will ask our legal team to to confirm that) will cover our needs.

**Comment by gberginc** [comment]
I agree that conserver could be an alternative, it's just that it brings a dependency that perhaps is not necessary. For me personally, the `virsh console` has very limited use so as a first step for us, it would suffice to redirect to a log file. VNC can still be used in case SSH is not available (which I think should be available in most regular OSes).

That said, I haven't used conserver yet so it may be trivial to hook it up.

**Comment by jellonek** [comment]
I think that we should have a both ways to connect to workload, So using something which is close to `kubectl logs` experience (e.x. using conserv, and possibly logging whole output of stdout by serio to some file) is a start.
The second would be something what is more coupled with existing VM interfaces, in libvirt meaning - for now we don't have any code in this direction.
If there will be a next possibility to conform k8s interfaces and still have an UI for virtual machines -  e.x. having a serio output as application log - this can be marked in docs as a possibility, but it's up to app developer if he want's to confirm to this.

btw. https://github.com/Mirantis/virtlet/pull/248 will provide a possibility to quite easy access to VM from host possibility, what means a quite easy access to vm console from cluster node. so we are working (and hopefuly we will agree) in similar areas :)

**Comment by miha-plesko** [comment]
Hello,
so I did some research about how to handle logging for Virtlet VMs. There were three approaches that I've researched:

1. use of Conserver program (suggested by @jellonek )
2. log serial console into file (suggested by @me)
3. redirect serial console into file disabling `virsh console` (suggested by @gberginc )


Turns out that (1) and (2) have some major drawbacks, so I went with (3) and implemented `virtlet_log` binary that I am kindly asking you to review ([see this PR](https://github.com/Mirantis/virtlet/pull/281)).

## 1 - use of Conserver program
I made an attempt to use Conserver program as middleware but I failed to think of an architecture. The main drawback of Conserver program are:

* it does not support dynamic addition of a new serial port to observe
* it does not support custom formatting of logs
* it appears not to work well with libvirt since it was meant to be used on actual HW
* it appears not to be maintained for 10 years now

The former bullet hurts the most. Intended usage of Conserver is:
```
prepare configuration file -> run Converver in background -> access it with (possibly multiple) clients
```
 However, when a new serial port appears (libvirt boots another VM), you need to stop the whole Conserver, update configuration file, and start Conserver again:
```
stop Conserver (and disconnect all connected clients!!!) ->
update configuration file ->
start Conserver in background ->
clients can reconnect
```
From Virtlet perspective where a lot of VM creations/deletions are expected, Conserver would be starting and stopping all the time disconnecting clients and possible loosing logs...

## 2 - log serial console into file
Libvirt 1.3.3 (or newer) supports [logging serial port into file](https://libvirt.org/formatdomain.html#elementsConsole) so that both features are supported: `virsh console` works and everything is written into file. Sure sounds great until you try it on your own: turns out that file is being written ONLY to log user's `virsh console` activity. So if `virsh console` is never used, the log file is empty... The drawbacks here are therefore quite impossible to workaround:

* only `virsh console` sessions are recorded into file (no grub messages...)

## 3 - redirect serial console into file
It was somewhat trivial to update libvirt XML to redirect serial console to file. Having this done, however, some other problems have arised:

* `virsh console` works no more
* Kubernetes expects to have logs in its own JSON format, not just plaintext dump

Luckily these two problems are possible to overcome.

+ Instead of `virsh console`, one can either SSH to the VM, or even better, one can VNC to the VM (that's what they use in OpenStack Horizon implementation).
+ Logs from file can be reformatted later on. I've already implemented a Go program to do this work and integrated it into Virtlet.


# Bottom line
There were two problems to solve - one is how we get logs out of VM, other is how we transform them into a form and location expected by Kubernetes. I did some research and [implemented ](https://github.com/Mirantis/virtlet/pull/281) the most reasonable solution I could think of. Please let me know if you have any suggestons how to improve it.

**Comment by jellonek** [comment]
Already implemented some time ago in #281- closing.

**Comment by justinc1** [comment]
@miha-plesko I remember talking a while ago about using serial console with rumprun (by default, only VGA is used in rumprun), and neither of us did know how to do this. Well, it is simple, per https://www.freelists.org/post/rumpkernel-users/Rumprun-serial-console-support-implemented-for-x86, just add '-nographic -vga none' to qemu cmdline - and it just works :)

Posting here, as it might be interesting to someone else too.
